### PR TITLE
Define `fn [_]::try_split_at(&self, usize) -> Option<(&Self, &Self)>`

### PR DESCRIPTION
**Pull request by strake**

### CODE REVIEW

File: src/libcore/slice/mod.rs [hunk 1]
```diff
@@ -804,6 +804,61 @@ impl<T> [T] {
...
+    ///
+    /// {
+    ///     assert!(v.split_at(7).is_none())
+    /// }
+    /// ```
+    #[unstable(feature = "try_split_at", issue = "54886")]
+    #[inline]
+    pub fn try_split_at(&self, mid: usize) -> Option<(&[T], &[T])> {
+        if mid > self.len() { None } else {
+            Some(unsafe { (self.get_unchecked(..mid), self.get_unchecked(mid..)) })
```
Comment (Centril) [inline src/libcore/slice/mod.rs]: Add a comment explaining why this is safe.

File: src/libcore/slice/mod.rs [hunk 2]
```diff
@@ -842,6 +897,43 @@ impl<T> [T] {
...
+    /// assert!(v == [1, 2, 3, 4, 5, 6]);
+    /// ```
+    #[unstable(feature = "try_split_at", issue = "54886")]
+    #[inline]
+    pub fn try_split_at_mut(&mut self, mid: usize) -> Option<(&mut [T], &mut [T])> {
+        let len = self.len();
+        let ptr = self.as_mut_ptr();
+
+        if mid > self.len() { None } else {
+            Some(unsafe { (from_raw_parts_mut(ptr, mid),
```
Comment (Centril) [inline src/libcore/slice/mod.rs]: Add a comment explaining why this is safe.

**Review by Centril — CHANGES_REQUESTED** [review]

File: src/libcore/slice/mod.rs (hunk 1 above)
Comment (rkruppe) [reply src/libcore/slice/mod.rs]: More importantly, please justify using unsafe at all. While it's possible that something `if mid > self.len() { None } else { (&self[..mid], &self[mid..]) }` may end up with redundant bounds checks, in principle they can be optimized out and we should only use unsafe code if that optimization fails to fire.

Comment (strake) [reply src/libcore/slice/mod.rs]: I just checked: it indeed elides the bounds checks and emits the same LLVM IR in either case.

**Comment by strake** [comment]
I redefined these functions sans `unsafe` blocks. PTAL

**Comment by scottmcm** [comment]
For my edification, can you elaborate why these are better than the caller just checking things?  Should there also be `try_` versions of other methods on slices that take indexes, like `try_swap` or `try_rotate_left`?  Since these methods are safe, couldn't they live in a crate instead of `core`?

**Comment by strake** [comment]
@scottmcm

I just think it's easier to read and harder to misuse in certain use cases — in my experience, for example: parsing certain binary formats, i want to return notice of invalid data if the given slice is too short, but if it is written in terms of `split_at` rather than `try_split_at` it is too easy to delete the bounds check in a later refactorization and not notice. I already defined this function in a utility crate but i thought others might also find it useful enough to include in core, rather than depend on yet another utility crate and `use` an extension trait, which is noise.

I seldom find myself checking bounds before calling `swap` or `rotate` and these methods return unit anyhow so are easier to wrap in a simple `if`.

**Comment by alexcrichton** [comment]
I personally agree with @scottmcm where I don't think we should add this to the standard library at this time. We don't want to have a dual of all the functionality on slices, both fallible and infallible. Instead we, long ago, decided that we'd only have either infallible or fallible and for slices it's already all related to indexing, which is always infallible, so slices inherited that behavior.

Considering to reverse this decision by adding `try_*` variants of all the various methods I believe is RFC-worthy rather than doing one-off in a PR. 

I agree that a local optimum can often be achieved with one extra method, but such a method can always be defined locally in a crate or as a library on crates.io.

**Comment by TimNN** [comment]
Ping from triage! Thank for your PR, @strake. Based on @alexcrichton's and @scottmcm's comments I'm going to close this PR as blocked on an RFC. Thanks for your contribution!

**Comment by Ben-Lichtman** [comment]
I know that this issue is long-closed, however I'd like to add that this would allow the removal of panicking code paths from code which may not panic, so I suggest a reconsideration.

**Pull request closed without merging**